File: server/domain/ports/text_typer.go
// screenreader-mcp domain -- the TextTyper port (the `typing` capability).
// Copyright (C) 2026 Marlon Brandao de Sousa. GPL-2. See COPYING.txt.
//
// ROLE: domain port, the `typing` capability group.
// IMPLEMENTED BY: adapters/bridge/json_lines_client.go.
// USED BY: the type_text tool controller.
// HANDED OUT BY: the handshake, only when the reader announced `typing`.
package ports

type TextTyper interface {
	// TypeText treats text as opaque; control characters go through
	// PressGestures. graceMs and announce behave as on PressGestures.
	TypeText(text string, graceMs int, announce string) (TypeOutcome, error)
}

File: bridges/nvda/tests/fakes/text_typer.py
# nvdaMcpBridge tests -- FakeTextTyper, standing in for the TextTyper port.
# Copyright (C) 2026 Marlon Brandao de Sousa. GPL-2. See COPYING.txt.

from __future__ import annotations

from collections.abc import Sequence

from nvdaMcpBridge.domain.ports.text_typer import TextTyper, TypingError


class FakeTextTyper(TextTyper):
	def __init__(self, *, fail_on: Sequence[str] | None = None) -> None:
		self._fail_on = set(fail_on or ())
		self.typed: list[str] = []

	def type_text(self, text: str) -> None:
		if text in self._fail_on:
			raise TypingError(f"typing failed: {text!r}")
		self.typed.append(text)
